lib/indexer.rb

```diff
@@ -0,0 +1,151 @@
+      index_log.level = ENV['LOG_LEVEL'] ? Logger::SEV_LABEL.index(ENV['LOG_LEVEL']) : Logger::INFO
+      $stdout.sync = true
+      start_time = Time.now
+
+
+      # TODO: break out pid list creation
+      $stdout.puts 'Discovering PIDs...'
+      index_log.info 'Discovering PIDs...'
+      dor_pids = []
+      solr_pids = []
+      if args[:query] != ':ALL:'
+        q = (args[:query].nil? || args[:query] == ':MISSING:') ? '*:*' : args[:query]
+        puts q
+        start = 0
+        resp = Dor::SearchService.query(q, :sort => 'id asc', :rows => 1000, :start => start, :fl => ['id'])
+        while resp.docs.length > 0
+          solr_pids += resp.docs.collect { |doc| doc['id'] }
+          start += 1000
+          $stdout.print '.'
+          resp = Dor::SearchService.query(q, :sort => 'id asc', :rows => 1000, :start => start, :fl => ['id'])
+        end
+        $stdout.puts
+        msg = "Found #{solr_pids.length} PIDs in solr."
+        $stdout.puts msg
+        index_log.info msg
+      end
+      if args[:query] =~ /:(ALL|MISSING):/
+        dor_pids = []
+        Dor::SearchService.iterate_over_pids(:in_groups_of => 1000, :mode => :group) do |chunk|
+          dor_pids += chunk
+          $stderr.print '.'
+        end
+        $stdout.puts
+        msg = "Found #{dor_pids.length} PIDs in DOR."
+        $stdout.puts msg
+        index_log.info msg
+      end
+      if dor_pids.present?
+        pids = dor_pids - solr_pids
+      else
+        pids = solr_pids
+      end
+      pids.delete_if { |pid| pid !~ /druid:/ }
+      time = Time.now - start_time
+      msg = "#{pids.length} PIDs discovered in #{[(time / 3600).floor, (time / 60 % 60).floor, (time % 60).floor].map{|tt| tt.to_s.rjust(2, '0')}.join(':')}"
+      $stdout.puts msg
+      index_log.info msg
+
+
+      # TODO: break out solrization of pid list
+      solr = ActiveFedora.solr.conn
+      pbar = ProgressBar.new('Reindexing...', pids.length)
+      errors = 0
+      pids.each do |pid|
+        begin
+          index_log.debug "Indexing #{pid}"
+          obj = Dor.load_instance pid
+          obj = obj.adapt_to(Dor::Item) if obj.class == ActiveFedora::Base
+          if obj.is_a?(Dor::Processable) && obj.workflows.new?
+            c = obj.class
+            obj.workflows.save
+            obj = Dor.load_instance(pid).adapt_to c
+          end
+          solr_doc = obj.to_solr
+          solr.add solr_doc, :add_attributes => {:commitWithin => 10}
+          errors = 0
+        rescue Interrupt
+          raise
+        rescue StandardError => e
+          errors += 1
+          index_log.warn("Error (#{errors}) indexing #{pid}")
+          index_log.error("#{e.class}: #{e.message}")
+          #  if errors == 3
+          #    index_log.fatal("Too many errors. Exiting.")
+          #    raise e
+          #  end
+        end
+        pbar.inc(1)
+      end # pids.each
+      solr.commit
+      puts 'Reindexing complete'
```

Comment:
since this is intended to be run from the command line, this doesn't seem like a misuse of stdout, unless there's a more appropriate way to write back to the user.  but it does seem like this would be categorized as standard output.